libraries/persistent_store/src/format.rs

```diff
@@ -0,0 +1,907 @@
+        } else if ID_REMOVE.check(word) {
+            let key = REMOVE_KEY.get(word);
+            ParsedWord::Internal(Internal::Remove { key })
+        } else if word == ERASED_WORD {
+            return Ok(WordState::Erased);
+        } else {
+            return Ok(WordState::Partial);
+        };
+        if let ParsedWord::Internal(internal) = &valid {
+            if WORD_CHECKSUM.get(word)? != 0 {
+                return Ok(WordState::Partial);
+            }
+            let invalid = match internal {
+                Internal::Erase { page } => *page > self.max_page(),
+                Internal::Clear { min_key } => *min_key > self.max_key(),
+                Internal::Marker { count } => *count > MAX_UPDATES,
+                Internal::Remove { key } => *key > self.max_key(),
+            };
+            if invalid {
+                return Err(StoreError::InvalidStorage);
+            }
+        }
+        Ok(WordState::Valid(valid))
+    }
+
+    /// Builds the storage representation of a user entry.
+    pub fn build_user(&self, key: usize, value: &[u8]) -> Vec<u8> {
+        let length = value.len();
+        let word_size = self.word_size();
+        let footer = self.bytes_to_words(length);
+        let mut result = vec![0xff; (1 + footer) * word_size];
+        result[word_size..][..length].copy_from_slice(value);
+        let mut word = ERASED_WORD;
+        ID_HEADER.set(&mut word);
+        if footer > 0 && is_erased(&result[footer * word_size..]) {
+            HEADER_FLIPPED.set(&mut word);
+            *result.last_mut().unwrap() = 0x7f;
+        }
+        HEADER_LENGTH.set(&mut word, length);
+        HEADER_KEY.set(&mut word, key);
+        HEADER_CHECKSUM.set(&mut word, count_zeros(&result[footer * word_size..]));
+        result[..word_size].copy_from_slice(&word.to_ne_bytes());
+        result
+    }
+
+    /// Sets the padding bit in the first word of a user entry.
+    ///
+    /// The word is taken as a slice for convenience.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `slice.len() != WORD_SIZE`.
+    pub fn set_padding(&self, slice: &mut [u8]) {
+        let mut word = slice_to_word(slice);
+        ID_PADDING.set(&mut word);
+        slice.copy_from_slice(&word.to_ne_bytes());
+    }
+
+    /// Sets the deleted bit in the first word of a user entry.
+    ///
+    /// The word is taken as a slice for convenience.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `slice.len() != WORD_SIZE`.
+    pub fn set_deleted(&self, slice: &mut [u8]) {
+        let mut word = slice_to_word(slice);
+        HEADER_DELETED.set(&mut word);
+        slice.copy_from_slice(&word.to_ne_bytes());
+    }
+
+    /// Returns the minimum number of words to represent a given number of bytes.
+    pub fn bytes_to_words(&self, bytes: usize) -> usize {
+        div_ceil(bytes, self.word_size())
+    }
+}
+
+/// The word index of the init info in a page.
+const INIT_WORD: usize = 0;
```

Comment:
Ok. Could be worth reminding that fact in documentation here.

Reply:
I don't see it done in the latest version of this file.

Reply:
It's in the documentation of `CONTENT_WORD` below because it's the highest word index. So if it doesn't overflow, neither do the smaller word indices. I don't think it's worth duplicating it to all of those.

Reply:
Ok. It could have been clearer before the first constant to find this information when reading from the top down, but fine.